target/linux/ramips/dts/mt7620a_dlink_dwr-960.dts

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include "mt7620a.dtsi"
+
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	compatible = "dlink,dwr-960", "ralink,mt7620a-soc";
+	model = "D-Link DWR-960";
+
+	aliases {
+		led-boot = &led_status;
+		led-failsafe = &led_status;
+		led-running = &led_status;
+		led-upgrade = &led_status;
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 1 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		wps {
+			label = "wps";
+			gpios = <&gpio0 2 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_status: status {
+			label = "dwr-960:green:status";
+			gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
+		};
+
+		wan {
+			label = "dwr-960:green:wan";
+			gpios = <&gpio0 11 GPIO_ACTIVE_LOW>;
+		};
+
+		lan {
+			label = "dwr-960:green:lan";
+			gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
+		};
+
+		sms {
+			label = "dwr-960:green:sms";
+			gpios = <&gpio1 15 GPIO_ACTIVE_LOW>;
+		};
+
+		signal_green {
+			label = "dwr-960:green:signal";
+			gpios = <&gpio2 0 GPIO_ACTIVE_HIGH>;
+		};
+
+		signal_red {
+			label = "dwr-960:red:signal";
+			gpios = <&gpio2 1 GPIO_ACTIVE_HIGH>;
+		};
+
+		4g {
+			label = "dwr-960:green:4g";
+			gpios = <&gpio2 2 GPIO_ACTIVE_LOW>;
+		};
+
+		3g {
+			label = "dwr-960:green:3g";
+			gpios = <&gpio2 3 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan5g {
+			label = "dwr-960:green:wlan5g";
+			gpios = <&gpio2 4 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+
+		wlan2g {
+			label = "dwr-960:green:wlan2g";
+			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy1tpt";
+		};
+	};
+};
+
+&ethernet {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&rgmii2_pins &mdio_pins>;
+	mediatek,portmap = "wllll";
+
+	port@5 {
+		status = "okay";
+		phy-mode = "rgmii-txid";
+		phy-handle = <&phy7>;
+	};
+
+	mdio-bus {
+		status = "okay";
+
+		phy7: ethernet-phy@7 {
+			reg = <7>;
+			phy-mode = "rgmii-id";
+		};
+
+	};
+};
+
+&gpio1 {
+	status = "okay";
+};
+
+&gpio2 {
+	status = "okay";
+};
+
+&gpio3 {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <80000000>;
```

Comment:
The frequency should be 50MHz, see https://github.com/openwrt/openwrt/commit/a508ab1ac843c5a60a134bf3cdc8837aee6b8ccc#commitcomment-37769002
It seems the slow flashing speed is common in mt7620 devices...

Reply:
Change to ```spi-max-frequency = <50000000>;``` causes unstable work and bootloops. I stay at 80M.

Reply:
Hmm, that's quite unexpected for me. Both 50MHz and 80MHz spi-max-frequency should return the same actual hardware frequency 48.3MHz. If you have time, could you please change this line from dev_dbg to dev_info and show me the boot log line?
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/patches-4.14/0042-SPI-ralink-add-Ralink-SoC-spi-driver.patch#L234

Reply:
I afraid, that in last case there was failed flash. In second try  with 50M everything was ok. I remember, some time ago I was the same situation with another d-link, sometime those flashes flash faultily.

I ran some tests. Output:
```
m25p80 spi0.0: speed: 48333333/80000000, rate: 4, prescal: 1, loops: 162
m25p80 spi0.0: speed: 48333333/50000000, rate: 4, prescal: 1, loops: 162
m25p80 spi0.0: speed: 24166666/48333333, rate: 8, prescal: 2, loops: 226
```

It looks, that everything >=48333334 is the same.

Reply:
Thanks for the test, let me know if it goes wrong again. I usually test the speed on initramfs, so it might act differently in firmware on real flash...

Reply:
FYI it seems the slow flashing speed is due to `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y`. After I removed it, `time mtd write /tmp/initramfs.bin firmware` decreased from 1m 34.65s to 22.20s.